torch/distributed/checkpoint/_fsspec_filesystem.py

```diff
@@ -33,8 +34,16 @@ def create_stream(
         self, path: Union[str, os.PathLike], mode: str
     ) -> Generator[io.IOBase, None, None]:
         assert self.fs is not None
-        with self.fs.transaction:
-            with fsspec.open(str(path), mode) as stream:
+
+        # fsspec transactions do not support concurrency and only allow
+        # 1 running transaction per filesystem. Because they do not
+        # guard against concurrent updates anyways, just create a new
+        # filesystem to isolate the transaction state. This is safe as
+        # long as we do not call `create_stream` concurrently on the
+        # same paths.
+        fs = copy.copy(self.fs)
```

Comment:
Does this actually solve the problem or does it need to be a deep copy?

Reply:
Not sure a shallow copy would be safe depending on how transactions are implemented on the backend for the specific fsspec endpoint.

Reply:
On second thought, I think this is likely safe only if the file system wasn't already in transaction from a 3rd party lib etc. We could add an assert on that boolean, but that wouldn't fully protect from the race condition.